### PR TITLE
Blocking method calls on Unknown

### PR DESCRIPTION
**Pull request by apanczyk**

### CODE REVIEW

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala [hunk 1]
```diff
@@ -289,7 +289,11 @@ private[spel] class Typer(classLoader: ClassLoader, commonSupertypeFinder: Commo
   }
 
   private def extractProperty(e: PropertyOrFieldReference, t: TypingResult): ValidatedNel[ExpressionParseError, TypingResult] = t match {
-    case Unknown => Valid(Unknown)
+    case Unknown =>
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala:292]: this is for the properties (e.g. fields), we also need check for methods - e.g. in TypeMethodReference

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala [hunk 2]
```diff
@@ -289,7 +289,11 @@ private[spel] class Typer(classLoader: ClassLoader, commonSupertypeFinder: Commo
...
+      if(disableMethodExecutionForUnknown)
+        invalid(s"There is no property '${e.getName}' in type: ${t.display}")
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala:294]: This message is not v. helpful as we already know that `t` is `Unknown` ;) Maybe sth like: "Property/method access on Unknown type is blocked"?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala (hunk 1 above)
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala:307]: Haven’t seen this, check for methods in TypeMethodReference is already implemented.

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala (hunk 2 above)
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala]: Yeah, you are right. Already changed that, like you said.

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala [hunk 3]
```diff
@@ -7,19 +7,22 @@ import pl.touk.nussknacker.engine.definition.TypeInfos.{ClazzDefinition, MethodI
...
+class TypeMethodReference(methodName: String, currentResults: List[TypingResult], calledParams: List[TypingResult], typeDefinitionSet: TypeDefinitionSet, disableMethodExecutionForUnknown: Boolean) {
   def call(implicit settings: ClassExtractionSettings): Either[String, TypingResult] =
     currentResults.headOption match {
       case Some(tc: SingleTypingResult) =>
         typeFromClazzDefinitions(extractClazzDefinitions(Set(tc)))
       case Some(TypedUnion(nestedTypes)) =>
         typeFromClazzDefinitions(extractClazzDefinitions(nestedTypes))
       case _ =>
-        Right(Unknown)
+        if(disableMethodExecutionForUnknown)
```
Comment (dswiecki) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala:22]: Can Unknown be returned from above cases ?

Comment (mproch) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala:22]: :+1: Good catch. Let's do the following:
- replace `currentResults: List[TypingResult]` with `invocationTarget: TypingResult`, as we already check that list is not empty in `Typer::extractMethodReference`
- add explicit `case Unknown` here

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala [hunk 4]
```diff
@@ -7,19 +7,22 @@ import pl.touk.nussknacker.engine.definition.TypeInfos.{ClazzDefinition, MethodI
...
+          Left("Method access on Unknown type is blocked")
```
[the remark this answers is not in the document]
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala]: Done

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala [hunk 5]
```diff
@@ -303,7 +304,11 @@ private[spel] class Typer(classLoader: ClassLoader, commonSupertypeFinder: Commo
   }
 
   private def extractProperty(e: PropertyOrFieldReference, t: TypingResult): ValidatedNel[ExpressionParseError, TypingResult] = t match {
-    case Unknown => Valid(Unknown)
+    case Unknown =>
+      if(disableMethodExecutionForUnknown)
+        invalid("Property access on Unknown type is blocked")
```
[the remark this answers is not in the document]
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala]: Changed

File: engine/interpreter/src/test/scala/pl/touk/nussknacker/engine/compile/ProcessValidatorSpec.scala [hunk 6]
```diff
@@ -93,11 +94,50 @@ class ProcessValidatorSpec extends FunSuite with Matchers with Inside {
...
+
+    val baseDefinitionCopy = baseDefinition.copy(
+      expressionConfig = baseDefinition.expressionConfig.copy(
+        disableMethodExecutionForUnknown = true))
+
+    val correctProcess = EspProcessBuilder
+      .id("process1")
+      .exceptionHandler()
+      .source("id1", "sourceWithUnknown")
+      .filter("filter1", "#input.imaginaryMethod")
```
[the remark this answers is not in the document]
Comment (apanczyk) [reply engine/interpreter/src/test/scala/pl/touk/nussknacker/engine/compile/ProcessValidatorSpec.scala]: Alright, already added.

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala]: Changed

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala (hunk 3 above)
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala]: Alright, things changed.

Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala]: Alright, things changed.

**Comment by apanczyk** [comment]
Request is ready for a review.

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala [hunk 7]
```diff
@@ -7,22 +7,20 @@ import pl.touk.nussknacker.engine.definition.TypeInfos.{ClazzDefinition, MethodI
...
+      case TypedUnion(nestedTypes) =>
         typeFromClazzDefinitions(extractClazzDefinitions(nestedTypes))
-      case _ =>
-        if(disableMethodExecutionForUnknown)
-          Left("Method access on Unknown type is blocked")
-        else
-          Right(Unknown)
+      case Unknown =>
+        if(disableMethodExecutionForUnknown) Left("Method invocation on Unknown is not allowed") else Right(Unknown)
+      case _ => Right(Unknown)
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala:23]: I think this is not needed ATM. There are no more cases - if there are, it should be catched by compiler

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala [hunk 8]
```diff
@@ -323,7 +323,7 @@ private[spel] class Typer(classLoader: ClassLoader, commonSupertypeFinder: Commo
     context.stack match {
       case _ :: tail =>
         typeChildren(validationContext, node, context.copy(stack = tail)) { typedParams =>
-          TypeMethodReference(reference.getName, context.stack, typedParams, typeDefinitionSet, disableMethodExecutionForUnknown) match {
+          TypeMethodReference(reference.getName, context.stack.head, typedParams, typeDefinitionSet, disableMethodExecutionForUnknown) match {
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala:326]: let's write:
`case head :: tail => ` two lines above, and here just `head` - .head/.get are always kind of a smell :)

**Review by mproch — COMMENTED** [review]
Almost there :)

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala (hunk 7 above)
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/typer/TypeMethodReference.scala]: Done

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala (hunk 8 above)
Comment (apanczyk) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/Typer.scala]: Ok, already improved

**Review by mproch — APPROVED** [review]
:+1:

**Pull request merged**